crypto/err/err.c

```diff
@@ -741,8 +741,19 @@ int err_shelve_state(void **state)
 {
     int saveerrno = get_last_sys_error();
 
+#if 0
+    /*
+     * It seems wrong to recurse into our only caller here, with options that
+     * may not match the outer call, and potentially cause failures when
+     * complementary run-once functions are skipped and return 0.  Nothing
+     * seems to require this call here, and tests pass without it.
+     *
+     * If there is some reason for this, a different solution may be needed.
+     */
```

Comment:
The INIT_BASE_ONLY calls make sure we allocate the pthread_keys in exactly the reverse order as they are
destroyed, see #5899 for details.

Of course since err_shelve_state is currently only called from crypto/init.c removing this call is harmless.
However if this function is called from elsewhere in the future this would not work without this call.

I think err_shelve_state is superior and could be used instead of ERR_set_mark/ERR_pop_to_mark
and would avoid issues with overflow if the error stack, which defeats ERR_pop_to_mark,
while err_shelve_state does not depend on the current error state.

Reply:
In the specific case of `err_shelve_state`, called only via `OPENSSL_init_crypto() -> ossl_init_load_crypto_nodelete()` we get I think wrong semantics by re-entering `OPENSSL_init_crypto()` with potentially incompatible option bits and `NULL` settings, before the outer call to `OPENSSL_init_crypto` is done.  So perhaps the solution is to add comments to `err_shelve_state()` that it MUST NOT be called before the base initialization in `OPENSSL_init_crypto` is done.  We could almost cut/paste the requisite code into `ossl_init_load_crypto_nodelete()`, except that `err_shelve_state()` calls `err_do_init()` which is static inside `err.c`.

Keep in mind also that running `OPENSSL_init_crypto()` a second time with any of the paired `OPENSSL_INIT_XXX/OPENSSL_INIT_NO_XXX` options flipped always "fails" (`return 0`).
I added a comment about this: <https://github.com/openssl/openssl/pull/7969/files#diff-a6179475cbd0e12e843d14372560ae41R592>.  Perhaps the pairs of functions should share not only their write once barrier but also their memoized result, so that when the initial call to one member of the pair succeeds, calls to the second member also succeed (without calling it).  That might resolve a bunch of potential problems, but we still need to avoid the re-entry via `err_shelve_state` in some way.

So I am looking for suggestions here. I think that the PR is currently at least not wrong (it seems you don't disagree), but I am not sure whether something better should be done here. The initialization code is a bit of a maze, so that's why I asked for three reviewers (more welcome, if someone feels inclined to dive into this code).

The key point is that presently, removing the call is not only harmless, but in fact necessary, in order to avoid potentially incorrect re-entry into `OPENSSL_init_crypto()` with incompatible options and settings.

I see that the `OPENSSL_INIT_BASE_ONLY` is only used in `err.c` in two functions that only want the thread locals initialized, but run all the way through the function.  Should `OPENSSL_init_crypto()` perhaps return early when called with `OPENSSL_INIT_BASE_ONLY`?  The name is rather misleading, since it actually initialized everything anyway as presently implemented.

Reply:
Yes, INIT_BASE_ONLY is meant to initialize just the minimum thread state, and return
immediately, with success, of course, it is internal, I think I did not export it to public
headers.  It should not look at any other flags.  So this issue should not arise,
or we have broken something.

Reply:
and, I think it may only be used alone, therefore all the other init functions are all skipped.

Reply:
Thanks, I see now that though it does not return early, all the other initialization functions have pre-conditions that require an appropriate option flag to be set, so the re-entry I observed may be sub-optimal, but perhaps not an outright problem.
To make this more clear, I'm inclined to add a `return 1` right after `ossl_init_base`:
```
    if (!RUN_ONCE(&base, ossl_init_base))
        return 0;
    if (opts & OPENSSL_INIT_BASE_ONLY)
        return 1;
```
Would that make sense?

Returning to `err_shelve_state()` then, what would you like to do there?  Perhaps leave the `NOOP` re-entry as-is, just it case there is some day a new caller that is not inside `OPENSSL_init_crypto()`?

Reply:
In fact all functions from err.c can be called before or while initializing,
in most cases even from the init once functions (with a few exceptions).
It is extremely painful to implement because we allow multi-threaded initializations :-(

Reply:
I put back the recursive call, and added commentary.  I think this is now ready.

I could also add text in the documentation of `CONF_modules_load_file()` that if `OPENSSL_init_crypto()` (perhaps via `OPENSSL_init_ssl()` was called with `OPENSSL_INIT_NO_LOAD_CONFIG`, then the built-in modules are not yet available, and a call to `OPENSSL_load_builtin_modules()` is needed in order for `CONF_modules_load_file()` to work.